### ISSUE TITLE
Scare quotes breaking link

### ISSUE THREAD

**Issue by jvoight** [LMFDB/lmfdb#2938]
On 
  http://beta.lmfdb.org/Genus2Curve/Q/2916/b/
(but not on
  http://www.lmfdb.org/Genus2Curve/Q/2916/b/
) there are scare quotes around "D_{2,3}", and this leads to a broken
  http://beta.lmfdb.org/SatoTateGroup/1.4.%22D_%7B3%2C2%7D%22
in place of
  http://www.lmfdb.org/SatoTateGroup/1.4.1.6.1b.

Where did the quotes come from?  (And they should match in and out `` and ''.)

**Comment by alexjbest** [comment]
This is a problem in the database, some groups have it some do not. Roughly 3000 rows have quotes. They are the ones which have ST groups with curly braces {} in.

It looks like this also breaks searching for curves by by Sato-Tate group.

**Comment by roed314** [comment]
It's possible that this came from https://github.com/LMFDB/lmfdb/blob/master/lmfdb/backend/encoding.py#L329
but I don't see why that code path would be triggered for a text column (recursing should be False).

**Comment by AndrewVSutherland** [comment]
This is not a problem in the database.  The data for the Sato-Tate groups is the same on both proddb and legendre (I just checked), and in both cases the name and pretty columns for Sato-Tate group 1.4.1.6.1b are the text string 'D_{3,2}', which is what it should be.

> sage: db.gps_sato_tate.lookup("1.4.1.6.1b")["pretty"]
> u'D_{3,2}'
> 

The code in the web branch handles this fine, but the code in the dev branch does not; it has to be caused by a recent code change.

**Comment by roed314** [comment]
You're looking in the wrong place in the database.  The problem was in `g2c_curves`, where the `st_group` column had extra quotes.  I just fixed it in place, and http://beta.lmfdb.org/Genus2Curve/Q/2916/b/ now works.  This data change should get pushed to prod.

**Comment by AndrewVSutherland** [comment]
The data on proddb is fine, and now I know what caused the problem.  I updated g2c_curves on Friday (as noted in https://github.com/LMFDB/lmfdb/issues/2886#issuecomment-480563584) using rewrite to populate the column geom_end_alg that I added.  The st_group column must have been changed as a side effect (even though I did not touch this column).

But it looks like you wiped out the column geom_end_alg that I added on Friday?  How exactly did you "fix" the st_group column?

I can easily recompute it, but before I do we should fix the problem in rewrite (if that is indeed where the problem is).  It should not be making changes in columns that were not changed, and if it is this could easily cause problems elsewhere.

**Comment by AndrewVSutherland** [comment]
I removed the critical label since this issue is no longer causing a problem on the website but I'll leave it open until we figure out/fix the problem that occurred when the table was updated during a rewrite.  I'm running a test now to verify that there is a problem in rewrite.

**Comment by roed314** [comment]
I ran the following SQL command:
```
lmfdb=> UPDATE g2c_curves SET st_group = regexp_replace(st_group, '"(.*)"', '\1');
UPDATE 66158
```
I can't imagine how this would affect the `geom_end_alg` column.

**Comment by AndrewVSutherland** [comment]
Never mind, geom_end_alg is still there, you didn't change it.

**Comment by AndrewVSutherland** [comment]
OK, rewrite is indeed the culprit here, and I found another issue.  To reproduce both problems do the following:

> sage: db.create_table_like("g2c_test","g2c_curves")
> Table g2c_test created in 0.215 secs
> sage: r=db.g2c_curves.lookup("2916.b.11664.1")
> sage: db.g2c_test.insert_many([r])
> ---------------------------------------------------------------------------
> DataError                                 Traceback (most recent call last)
> <ipython-input-42-f60c5a3ae980> in <module>()
> ----> 1 db.g2c_test.insert_many([r])
> 
> /home/drew/github/lmfdb/lmfdb/backend/database.pyc in insert_many(self, data, resort, reindex, restat, commit)
>    2614                 inserter = inserter.format(Identifier(table),
>    2615                                            SQL(", ").join(map(Identifier, L[0].keys())))
> -> 2616                 self._execute(inserter, L, values_list=True, template=template)
>    2617             print "Inserted %s records into %s in %.3f secs"%(len(search_data), self.search_table, time.time()-now)
>    2618             self._break_order()
> 
> /home/drew/github/lmfdb/lmfdb/backend/database.pyc in _execute(self, query, values, silent, values_list, template, commit, slow_note, reissued)
>     358                 if template is not None:
>     359                     template = template.as_string(self.conn)
> --> 360                 execute_values(cur, query.as_string(self.conn), values, template)
>     361             else:
>     362                 try:
> 
> /scratch/sage/sage-8.3/local/lib/python2.7/site-packages/psycopg2/extras.pyc in execute_values(cur, sql, argslist, template, page_size)
>    1275             parts.append(b',')
>    1276         parts[-1:] = post
> -> 1277         cur.execute(b''.join(parts))
>    1278 
>    1279 
> 
> DataError: malformed array literal: "6.0.186624.1"
> LINE 1: ...e, '[[2,[1,0,2]],[3,[1]]]', true, '2916.b', ARRAY['6.0.18662...
>                                                              ^
> DETAIL:  Array value must start with "{" or dimension information.
> 
> sage: r["two_torsion_field"]=None
> sage: db.g2c_test.insert_many([r])
> Inserted 1 records into g2c_test in 0.002 secs
> resorting disabled
> sage: db.g2c_test.rewrite(lambda x: x)
> Reloading g2c_test...
> 	Loaded data into g2c_test in 0.094 secs from /tmp/tmprtKZaL
> Built primary key on g2c_test in 0.033 secs
> Swapped temporary tables for g2c_test into place in 0.0293328762054 secs
> New backup at g2c_test_old1
> Reloaded g2c_test in 0.269 secs
> sage: db.g2c_test.lookup("2916.b.11664.1")
> {u'Lhash': u'90343978874990028',
>  u'abs_disc': 11664L,
>  u'analytic_rank': 0,
>  u'aut_grp_id': u'[4,2]',
>  u'bad_lfactors': u'[[2,[1,0,2]],[3,[1]]]',
>  u'class': u'2916.b',
>  u'cond': 2916L,
>  u'disc_sign': -1,
>  u'eqn': u'[[0,0,0,0,0,0,1],[1]]',
>  u'g2_inv': u"['6400000/3','440000/9','-32000/81']",
>  u'geom_aut_grp_id': u'[24,8]',
>  u'geom_end_alg': u'M_2(CM)',
>  u'globally_solvable': 1,
>  u'has_square_sha': True,
>  u'igusa_clebsch_inv': u"['-960','25920','-7672320','-47775744']",
>  u'igusa_inv': u"['-120','330','320','-36825','-11664']",
>  u'is_gl2_type': True,
>  u'is_simple_base': False,
>  u'is_simple_geom': False,
>  u'label': u'2916.b.11664.1',
>  u'locally_solvable': True,
>  u'non_solvable_places': {},
>  u'num_rat_pts': 4,
>  u'num_rat_wpts': 0,
>  u'real_geom_end_alg': u'M_2(C)',
>  u'root_number': 1,
>  u'st_group': u'"D_{3,2}"',
>  u'torsion_order': 9,
>  u'torsion_subgroup': u'[3,3]',
>  u'two_selmer_rank': 0}
> 
> Note the square quotes that have been added were not in the original
> 
> sage: db.create_table_like("g2c_test","g2c_curves")
> Table g2c_test created in 0.215 secs
> sage: r=db.g2c_curves.lookup("2916.b.11664.1")
> sage: db.g2c_test.insert_many([r])
> ---------------------------------------------------------------------------
> DataError                                 Traceback (most recent call last)
> <ipython-input-42-f60c5a3ae980> in <module>()
> ----> 1 db.g2c_test.insert_many([r])
> 
> /home/drew/github/lmfdb/lmfdb/backend/database.pyc in insert_many(self, data, resort, reindex, restat, commit)
>    2614                 inserter = inserter.format(Identifier(table),
>    2615                                            SQL(", ").join(map(Identifier, L[0].keys())))
> -> 2616                 self._execute(inserter, L, values_list=True, template=template)
>    2617             print "Inserted %s records into %s in %.3f secs"%(len(search_data), self.search_table, time.time()-now)
>    2618             self._break_order()
> 
> /home/drew/github/lmfdb/lmfdb/backend/database.pyc in _execute(self, query, values, silent, values_list, template, commit, slow_note, reissued)
>     358                 if template is not None:
>     359                     template = template.as_string(self.conn)
> --> 360                 execute_values(cur, query.as_string(self.conn), values, template)
>     361             else:
>     362                 try:
> 
> /scratch/sage/sage-8.3/local/lib/python2.7/site-packages/psycopg2/extras.pyc in execute_values(cur, sql, argslist, template, page_size)
>    1275             parts.append(b',')
>    1276         parts[-1:] = post
> -> 1277         cur.execute(b''.join(parts))
>    1278 
>    1279 
> 
> DataError: malformed array literal: "6.0.186624.1"
> LINE 1: ...e, '[[2,[1,0,2]],[3,[1]]]', true, '2916.b', ARRAY['6.0.18662...
>                                                              ^
> DETAIL:  Array value must start with "{" or dimension information.
> 
> sage: r["two_torsion_field"]=None
> sage: db.g2c_test.insert_many([r])
> Inserted 1 records into g2c_test in 0.002 secs
> resorting disabled
> sage: db.g2c_test.rewrite(lambda x: x)
> Reloading g2c_test...
> 	Loaded data into g2c_test in 0.094 secs from /tmp/tmprtKZaL
> Built primary key on g2c_test in 0.033 secs
> Swapped temporary tables for g2c_test into place in 0.0293328762054 secs
> New backup at g2c_test_old1
> Reloaded g2c_test in 0.269 secs
> sage: db.g2c_test.lookup("2916.b.11664.1")
> {u'Lhash': u'90343978874990028',
>  u'abs_disc': 11664L,
>  u'analytic_rank': 0,
>  u'aut_grp_id': u'[4,2]',
>  u'bad_lfactors': u'[[2,[1,0,2]],[3,[1]]]',
>  u'class': u'2916.b',
>  u'cond': 2916L,
>  u'disc_sign': -1,
>  u'eqn': u'[[0,0,0,0,0,0,1],[1]]',
>  u'g2_inv': u"['6400000/3','440000/9','-32000/81']",
>  u'geom_aut_grp_id': u'[24,8]',
>  u'geom_end_alg': u'M_2(CM)',
>  u'globally_solvable': 1,
>  u'has_square_sha': True,
>  u'igusa_clebsch_inv': u"['-960','25920','-7672320','-47775744']",
>  u'igusa_inv': u"['-120','330','320','-36825','-11664']",
>  u'is_gl2_type': True,
>  u'is_simple_base': False,
>  u'is_simple_geom': False,
>  u'label': u'2916.b.11664.1',
>  u'locally_solvable': True,
>  u'non_solvable_places': {},
>  u'num_rat_pts': 4,
>  u'num_rat_wpts': 0,
>  u'real_geom_end_alg': u'M_2(C)',
>  u'root_number': 1,
>  u'st_group': u'"D_{3,2}"',
>  u'torsion_order': 9,
>  u'torsion_subgroup': u'[3,3]',
>  u'two_selmer_rank': 0}
> 

Note that the square quotes that have been added were not in the original:

> sage: db.g2c_curves.lookup("2916.b.11664.1")
> {u'Lhash': u'90343978874990028',
>  u'abs_disc': 11664L,
>  u'analytic_rank': 0,
>  u'aut_grp_id': u'[4,2]',
>  u'bad_lfactors': u'[[2,[1,0,2]],[3,[1]]]',
>  u'class': u'2916.b',
>  u'cond': 2916L,
>  u'disc_sign': -1,
>  u'eqn': u'[[0,0,0,0,0,0,1],[1]]',
>  u'g2_inv': u"['6400000/3','440000/9','-32000/81']",
>  u'geom_aut_grp_id': u'[24,8]',
>  u'geom_end_alg': u'M_2(CM)',
>  u'globally_solvable': 1,
>  u'has_square_sha': True,
>  u'igusa_clebsch_inv': u"['-960','25920','-7672320','-47775744']",
>  u'igusa_inv': u"['-120','330','320','-36825','-11664']",
>  u'is_gl2_type': True,
>  u'is_simple_base': False,
>  u'is_simple_geom': False,
>  u'label': u'2916.b.11664.1',
>  u'locally_solvable': True,
>  u'non_solvable_places': [],
>  u'num_rat_pts': 4,
>  u'num_rat_wpts': 0,
>  u'real_geom_end_alg': u'M_2(C)',
>  u'root_number': 1,
>  u'st_group': u'D_{3,2}',
>  u'torsion_order': 9,
>  u'torsion_subgroup': u'[3,3]',
>  u'two_selmer_rank': 0,
>  u'two_torsion_field': [u'6.0.186624.1', [2, 0, 0, -2, 0, 0, 1], [6, 3]]}

**Comment by AndrewVSutherland** [comment]
Just to be clear (github managed to garble the quote), if you try to insert_many the record returned by db.g2c_test.lookup("2916.b.11664.1") it fails complaining about the two_torsion_field (a jsonb), and doing a rewrite with the identity function modifies st_group (adding scare quotes).

**Comment by roed314** [comment]
I think I fixed the scare quotes in #2941.  I think the insert_many is a separate bug which I'll need to look into.

**Comment by AndrewVSutherland** [comment]
Great! I’ll test the scare quote fix shortly.

**Comment by AndrewVSutherland** [comment]
Fixed by #2941
